algorithms/src/snark/marlin/marlin.rs

```diff
@@ -423,8 +428,12 @@ where
         // --------------------------------------------------------------------
         // Third round
 
-        let (prover_third_message, third_oracles, prover_state) =
-            AHPForR1CS::<_, MM>::prover_third_round(&verifier_second_msg, prover_state, zk_rng)?;
+        let (prover_third_message, third_oracles, prover_state) = AHPForR1CS::<_, MM>::prover_third_round(
+            &verifier_first_message,
```

Comment:
We should just store `verifier_first_message` in `prover state`, so that we don't have to pass it explicitly.

Also, it seems worth while to make `prover_xyz_round` methods into methods on `ProverState`, so that we can call them like

`let (prover_third_message, third_oracles) = prover_state.third_round(...)`

(The round functions mutate `state` as they go along).

Similarly for `VerifierState`.

Reply:
Good idea, will look a lot cleaner. I will also think about how the AHP and State objects should cleanly relate. I started on this but this will be another massive diff and doesn't require any cryptography review, so I suggest I fix this in a future PR.

Reply:
https://github.com/AleoHQ/snarkVM/pull/1875